src/Storages/StorageS3.cpp

```diff
@@ -562,6 +574,142 @@ size_t StorageS3Source::ReadTaskIterator::estimatedKeysCount()
     return buffer.size();
 }
 
+
+StorageS3Source::ArchiveIterator::ArchiveIterator(
+    std::unique_ptr<IIterator> basic_iterator_,
+    const std::string & archive_pattern_,
+    const Configuration & configuration_,
+    ContextPtr context_,
+    KeysWithInfo * read_keys_)
+    : WithContext(context_)
+    , basic_iterator(std::move(basic_iterator_))
+    , basic_key_with_info_ptr(nullptr)
+    , configuration(configuration_)
+    , read_keys(read_keys_)
+{
+    if (archive_pattern_.find_first_of("*?{") != std::string::npos)
+    {
+        auto matcher = std::make_shared<re2::RE2>(makeRegexpPatternFromGlobs(archive_pattern_));
+        if (!matcher->ok())
+            throw Exception(
+                ErrorCodes::CANNOT_COMPILE_REGEXP, "Cannot compile regex from glob ({}): {}", archive_pattern_, matcher->error());
+        filter = IArchiveReader::NameFilter{[matcher, matcher_mutex = std::make_shared<std::mutex>()](const std::string & p) mutable
+                                            {
+                                                std::lock_guard lock(*matcher_mutex);
+                                                return re2::RE2::FullMatch(p, *matcher);
+                                            }};
+    }
+    else
+    {
+        path_in_archive = archive_pattern_;
+    }
+}
+
+void StorageS3Source::ArchiveIterator::lazyInitializeSafe()
+{
+    if (!initialized)
+    {
+        basic_key_with_info_ptr = basic_iterator->next();
+        refreshArchive();
+        initialized = true;
+    }
+}
+
+KeyWithInfoPtr StorageS3Source::ArchiveIterator::next(size_t)
+{
+    if (!path_in_archive.empty())
+    {
+        while (true)
+        {
+            std::unique_lock lock{take_next_mutex};
+            lazyInitializeSafe();
+            if (!archive_reader)
+                return {};
+            KeyWithInfoPtr current_basic_key_with_info_ptr = basic_key_with_info_ptr;
+            std::shared_ptr<IArchiveReader> current_archive_reader = archive_reader;
+            basic_key_with_info_ptr = basic_iterator->next();
+            refreshArchive();
```

Comment:
This part looks a bit odd. Does not look good that we do `next()` to get the next file (which also fetches object metadata and might result in additional list request as listing is done in batches), then do `refreshArchive` and initialize reader for the next file, _while we still haven't processed previous file_ (imagine that a query has a `LIMIT` clause, this might mean that we do not need to process all the files). I'd suggest to do something like this - remove `lazyInitializeSafe` and `refreshArchive` methods - no need in them as we do everything in the scope of one method - so `basic_key_with_info_ptr` is initialized first only inside the loop:
```
while (true)
{
    std::unique_lock lock{take_next_mutex};

    basic_key_with_info = basic_iterator->next();
    if (!basic_key_with_info)
        return {};

    archive_reader = createArchiveReader(...);
    bool file_exists = archive_reader->fileExists(path_in_archive);
    if (file_exists)
    {
         ...
    }
}
```
The loop could be simplified as well to something like: 
```
while (true)
{
    std::unique_lock lock{take_next_mutex};
    if (!file_enumerator)
    {
        basic_key_with_info = basic_iterator->next();
        if (!basic_key_with_info)
            return {};

        archive_reader = createArchiveReader(...);
        file_enumerator = archive_reader->firstFile();
    }
    if (!file_enumerator->nextFile())
    {
        file_enumerator.reset();
        continue;
    }

    String current_filename = file_enumerator->getFileName();
    ... process current file ...
}

```